File: src/sorting/almost_sorted.py
import os
import sys
sys.path.append(os.path.join(os.path.dirname(__file__), '../tools'))

import files

from enum import Enum


'''
    submitted code:

    def almost_sorted(array):
        increasing = []
        decreasing = []

        def fits(start, end):
            return array[end - 1] <= array[start] <= array[end + 1]

        for i in range(1, len(array) - 1):
            if array[i - 1] < array[i] > array[i + 1]:
                increasing.append(i)
            if array[i - 1] > array[i] < array[i + 1]:
                decreasing.append(i)

        if len(increasing) == 0 and len(decreasing) == 0:
            return Sorted.yes, None

        if len(increasing) == 1 and len(decreasing) == 1:
            if fits(increasing[0], decreasing[0]) and fits(decreasing[0], increasing[0]):
                return Sorted.swap if decreasing[0] - increasing[0] <= 2 else Sorted.reverse, (increasing[0], decreasing[0])

        if len(increasing) == 2 and len(decreasing) == 2:
            if fits(increasing[0], decreasing[1]) and fits(decreasing[1], increasing[0]):
                return Sorted.swap, (increasing[0], decreasing[1])

        return Sorted.no, None


    def main():
        n     = int(input())
        d     = [-1] + [int(i) for i in input().split()] + [1000001]
        s, p  = almost_sorted(d)

        if s is not Sorted.no:
            print('yes')
            if p:
                print('%s %s %s' % (s.name, p[0], p[1]))
        else:
            print('no')


    if __name__ == "__main__":
        main()

'''

Sorted = Enum('Sorted', 'yes reverse swap no')


def almost_sorted(array):
    increasing = []
    decreasing = []

    def fits(start, end):
        return array[end - 1] <= array[start] <= array[end + 1]

    for i in range(1, len(array) - 1):
        if array[i - 1] < array[i] > array[i + 1]:
            increasing.append(i)
        if array[i - 1] > array[i] < array[i + 1]:
            decreasing.append(i)

    if len(increasing) == 0 and len(decreasing) == 0:
        return Sorted.yes, None

    if len(increasing) == 1 and len(decreasing) == 1:
        if fits(increasing[0], decreasing[0]) and fits(decreasing[0], increasing[0]):
            return Sorted.swap if decreasing[0] - increasing[0] <= 2 else Sorted.reverse, (increasing[0], decreasing[0])

    if len(increasing) == 2 and len(decreasing) == 2:
        if fits(increasing[0], decreasing[1]) and fits(decreasing[1], increasing[0]):
            return Sorted.swap, (increasing[0], decreasing[1])

    return Sorted.no, None


def main(argv):
    lines = files.read_lines_of_ints(argv[0])
    n     = lines[0][0]
    d     = [-1] + lines[1] + [1000001]
    s, p  = almost_sorted(d)

    if s is not Sorted.no:
        print('yes')
        if p:
            print('%s %s %s' % (s.name, p[0], p[1]))
    else:
        print('no')


if __name__ == "__main__":
    main(sys.argv[1:])

File: src/sorting/almost_sorted.cpp
#include <vector>
#include <iostream>

using namespace std;


/*
    [ jerry@pantagruel hackerrank ] $ clang++ -O3 -msse2 -o almost_sorted src/sorting/almost_sorted.cpp

    [ jerry@pantagruel hackerrank ] $ time ./almost_sorted < data/sorting/almost_sorted_01.txt 

    yes
    swap 1 2

    real    0m0.005s
    user    0m0.001s
    sys     0m0.002s

*/


bool fits(vector<int> & d, int start, int end) {

    return d[end - 1] <= d[start] && d[start] <= d[end + 1];
}


int main() {

    int n;
    cin >> n;

    vector<int> d(n + 2);
    d[0] = -1;
    for(int i = 1; i <= n; i++)
        cin >> d[i];
    d[n + 1] = 1000001;

    int increasing = 0;
    int decreasing = 0;

    int iindex     = 0;
    int dindex     = 0;

    for (int i = 1; i < n + 1; i++) {

        if (d[i - 1] < d[i] && d[i] > d[i + 1]) {

            increasing++;
            if (iindex == 0)
                iindex = i;
        }

        if (d[i - 1] > d[i] && d[i] < d[i + 1]) {

            decreasing++;
            dindex = i;
        }
    }

    if (increasing == 0 && decreasing == 0) {

        cout << "yes" << endl;
        return 0;
    } 

    if (increasing == 1 && decreasing == 1) {

        if (fits(d, iindex, dindex) && fits(d, dindex, iindex)) {

            if (dindex - iindex <= 2)
                cout << "yes\nswap" << " " <<  iindex << " " << dindex << endl;
            else
                cout << "yes\nreverse" << " " <<  iindex << " " << dindex << endl;
            return 0;
        }
    }

    if (increasing == 2 && decreasing == 2) {

        if (fits(d, iindex, dindex) && fits(d, dindex, iindex)) {

            cout << "yes\nswap" << " " <<  iindex << " " << dindex << endl;
            return 0;
        }
    }

    cout << "no" << endl;
}
